### PR TITLE
extend bucket_internal to SAMPLE generation mode

### PR DESCRIPTION
**Pull request by xt574chen**
# What does this PR do?

Extend function #24 to sample mode.
image

The command to reproduce performance is as follows:
`python ../gaudi_spawn.py --use_deepspeed --world_size 4 run_generation.py --model_name_or_path meta-llama/Llama-2-70b-hf --use_hpu_graphs --use_kv_cache --max_input_tokens 128 --max_new_tokens 2048 --batch_size 240 --attn_softmax_bf16  --trim_logits --bf16 --reuse_cache --warmup 1 --n_iterations 1 --limit_hpu_graphs --do_sample --bucket_size 256 --bucket_internal`

### CODE REVIEW

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -735,7 +735,7 @@ def generate(
         # if generation_config.bucket_size <= 0, padding is handled by the generating fn (like greedy_search)
         if generation_config.static_shapes and generation_config.bucket_size > 0:
             assert (
-                generation_mode == GenerationMode.GREEDY_SEARCH or generation_mode == GenerationMode.BEAM_SEARCH
+                generation_mode == GenerationMode.GREEDY_SEARCH or generation_mode == GenerationMode.SAMPLE
```
Comment (puneeshkhanna) [inline optimum/habana/transformers/generation/utils.py:738]: Lets have the check of BEAM SEARCH too since bucketing changes from Sayantan works in beam search.

File: optimum/habana/transformers/generation/utils.py [hunk 2]
```diff
@@ -1760,6 +1763,19 @@ def sample(
             model_kwargs["lazy_mode"] = lazy_mode
             model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
 
+            if bucket_size > 0 and bucket_internal:
```
Comment (puneeshkhanna) [inline optimum/habana/transformers/generation/utils.py:1766]: These changes should be after model fwd call.

File: optimum/habana/transformers/generation/utils.py [hunk 3]
```diff
@@ -1825,6 +1846,18 @@ def sample(
...
             )
+            if bucket_size > 0 and bucket_internal:
+                # Calculate slice idx for kv cache during the decode phase.
+                # Breaking down the kv cache in the attention block helps to reduce computation time.
+                if model_kwargs.get("token_idx_cpu") <= (model_kwargs["kv_cache_len"] // bucket_size) * bucket_size:
+                    idx = (model_kwargs.get("token_idx_cpu") - 1) // bucket_size
+                    if prev_idx != idx:
+                        model_kwargs["cache_idx"] = (idx + 1) * bucket_size
+                        prev_idx = idx
+                        if model_kwargs["use_hpu_graphs"]:
```
Comment (puneeshkhanna) [inline optimum/habana/transformers/generation/utils.py:1857]: remove these 2 lines for clear cache

**Comment by puneeshkhanna** [comment]
@dvarshney-habana  - Changes look good to me. We can merge so that sampling search starts working with bucket_internal.

**Review by dvarshney-habana — APPROVED** [review]
